rust/blockstore/src/arrow/sparse_index.rs

```diff
@@ -361,213 +361,104 @@ impl SparseIndexReader {
-                    SparseIndexDelimiter::Key(k) => Some(k),
-                };
-                // If delimeter starts with the same prefix then there will be keys inside the
-                // block with this prefix.
-                if non_start_curr_key.is_some()
-                    && prefix == non_start_curr_key.unwrap().prefix.as_str()
-                {
-                    block_ids.push(curr_block_value.id);
-                }
-                // If prefix is between the current delim and next delim then there could
-                // be keys in this block that have this prefix.
-                if (non_start_curr_key.is_none()
-                    || prefix > non_start_curr_key.unwrap().prefix.as_str())
-                    && (prefix <= non_start_next_key.unwrap().prefix.as_str())
-                {
-                    block_ids.push(curr_block_value.id);
-                }
-            } else {
-                // Last block.
-                if non_start_curr_key.is_none()
-                    || prefix >= non_start_curr_key.unwrap().prefix.as_str()
-                {
-                    block_ids.push(curr_block_value.id);
-                }
-            }
-        }
-        block_ids
-    }
-
-    /// Get all block ids that have keys with the given prefix and key greater than the given key
-    pub(super) fn get_block_ids_gt<'a, K: ArrowReadableKey<'a> + Into<KeyWrapper>>(
+    pub(super) fn get_block_ids_range<
+        'prefix,
+        'a,
+        K: ArrowReadableKey<'a> + Into<KeyWrapper>,
+        PrefixRange,
+        KeyRange,
+    >(
         &self,
-        prefix: &str,
-        key: K,
-    ) -> Vec<Uuid> {
-        let data = &self.data;
-        let forward = &data.forward;
-        let curr_iter = forward.iter();
-        let mut next_iter = forward.iter().skip(1);
-        let mut block_ids = vec![];
-        for (curr_delim, curr_block_value) in curr_iter {
-            let curr_key = match curr_delim {
-                SparseIndexDelimiter::Start => None,
-                SparseIndexDelimiter::Key(k) => Some(k),
-            };
-            let mut next_key: Option<&CompositeKey> = None;
-            if let Some((next_delim, _)) = next_iter.next() {
-                next_key = match next_delim {
-                    SparseIndexDelimiter::Start => {
-                        panic!("Invariant violation. Sparse index is not valid.")
-                    }
-                    SparseIndexDelimiter::Key(k) => Some(k),
+        // These key ranges are flattened instead of using a single RangeBounds<CompositeKey> because not all keys have a well-defined min and max value. E.x. if the key is a string, there would be no way to get the range for all keys within a specific prefix.
+        prefix_range: PrefixRange,
+        key_range: KeyRange,
+    ) -> Vec<Uuid>
+    where
+        PrefixRange: RangeBounds<&'prefix str>,
+        KeyRange: RangeBounds<K>,
+    {
+        let forward = &self.data.forward;
```

Comment:
Can we just use the range api?

Reply:
no, for a similar reason to my comment above
if only a prefix range is specified and the key range is unbounded, then there is no way to construct a `CompositeKey` for search within the tree.
we can potentially speed this up in the future with binary search if it proves to be a bottleneck